File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamfriends/callback/FriendMsgHistoryCallback.java
package in.dragonbra.javasteam.steam.handlers.steamfriends.callback;

import in.dragonbra.javasteam.enums.EAccountType;
import in.dragonbra.javasteam.enums.EResult;
import in.dragonbra.javasteam.enums.EUniverse;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientChatGetFriendMessageHistoryResponse;
import in.dragonbra.javasteam.steam.handlers.steamfriends.FriendMessage;
import in.dragonbra.javasteam.steam.handlers.steamfriends.SteamFriends;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.SteamID;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

/**
 * This callback is fired in response to receiving historical messages.
 * @see SteamFriends#requestOfflineMessages()
 */
public class FriendMsgHistoryCallback extends CallbackMsg {
    private EResult result;

    private SteamID steamID;

    private List<FriendMessage> messages;

    public FriendMsgHistoryCallback(CMsgClientChatGetFriendMessageHistoryResponse.Builder msg, EUniverse universe) {
        result = EResult.from(msg.getSuccess());

        steamID = new SteamID(msg.getSteamid());

        List<FriendMessage> messages = new ArrayList<>();
        for (CMsgClientChatGetFriendMessageHistoryResponse.FriendMessage m : msg.getMessagesList()) {
            SteamID senderID = new SteamID(m.getAccountid(), universe, EAccountType.Individual);
            Date timestamp = new Date(m.getTimestamp() * 1000L);

            messages.add(new FriendMessage(senderID, m.getUnread(), m.getMessage(), timestamp));
        }

        this.messages = Collections.unmodifiableList(messages);
    }

    public EResult getResult() {
        return result;
    }

    public SteamID getSteamID() {
        return steamID;
    }

    public List<FriendMessage> getMessages() {
        return messages;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamfriends/callback/FriendMsgHistoryCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamfriends.callback

import `in`.dragonbra.javasteam.enums.EAccountType
import `in`.dragonbra.javasteam.enums.EResult
import `in`.dragonbra.javasteam.enums.EUniverse
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientChatGetFriendMessageHistoryResponse
import `in`.dragonbra.javasteam.steam.handlers.steamfriends.FriendMessage
import `in`.dragonbra.javasteam.steam.handlers.steamfriends.SteamFriends
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg
import `in`.dragonbra.javasteam.types.SteamID
import java.util.*

/**
 * This callback is fired in response to receiving historical messages.
 *
 * @see SteamFriends.requestOfflineMessages
 * @see SteamFriends.requestMessageHistory
 */
class FriendMsgHistoryCallback(
    msg: CMsgClientChatGetFriendMessageHistoryResponse.Builder,
    universe: EUniverse?,
) : CallbackMsg() {

    /**
     * Gets the result of the request.
     */
    val result: EResult = EResult.from(msg.success)

    /**
     * Gets the [SteamID] of the user with whom these messages were exchanged.
     */
    val steamID: SteamID = SteamID(msg.steamid)

    /**
     * The messages exchanged with the user.
     * Offline messages are marked by having set [FriendMessage.unread] to **true**
     */
    val messages: List<FriendMessage> = msg.messagesList.map { m ->
        val senderID = SteamID(m.accountid.toLong(), universe, EAccountType.Individual)
        val timestamp = Date(m.timestamp * 1000L)
        FriendMessage(senderID, m.unread, m.message, timestamp)
    }
}
